plugin/dnstap/handler.go

```diff
@@ -15,20 +17,42 @@ import (
 type Dnstap struct {
 	Next plugin.Handler
 	io   tapper
+	repl replacer.Replacer
 
 	// IncludeRawMessage will include the raw DNS message into the dnstap messages if true.
 	IncludeRawMessage bool
 	Identity          []byte
 	Version           []byte
+	ExtraFormat       string
 }
 
-// TapMessage sends the message m to the dnstap interface.
+// TapMessage sends the message m to the dnstap interface, without interpreting the "Extra" field using metadata.
 func (h Dnstap) TapMessage(m *tap.Message) {
+	h.TapMessageWithMetadata(nil, m, request.Request{})
+}
+
+// TapMessageWithMetadata sends the message m to the dnstap interface, with "Extra" field being interpreted by the provided metadata context.
+func (h Dnstap) TapMessageWithMetadata(ctx context.Context, m *tap.Message, state request.Request) {
 	t := tap.Dnstap_MESSAGE
-	h.io.Dnstap(&tap.Dnstap{Type: &t, Message: m, Identity: h.Identity, Version: h.Version})
+	extraStr := h.ExtraFormat
+	if ctx != nil {
```

Comment:
please move this `if` under `if extraStr != ""` below. We don't want spending resources on Replacer if `extra` was not configured. So, only the users who configured `extra` will pay for it.
Because I see this method is called now even from internal functions like `tapQuery()`